### PR TITLE
Adds dog AI, catches up to COD: Ghosts

### PR DESCRIPTION
**Pull request by Ryll-Ryll**
## About The Pull Request
This PR uses Floyd's AI datums to give dogs some wonderful new behavior that I think people will appreciate. Namely, any time a dog sees someone throw something that lands within the dog's view, the dog will try running after it to grab and bring back to the thrower. The dog doesn't really care what it was you just threw, whether it was a bouncy toy ball or a live hand grenade, [as shown in this linked video.](https://streamable.com/6mbv9i) I'll probably put some limitations on fetching, like they can only carry up to a certain size item, but I'll finish that up last.

This PR also lets you befriend dogs by either petting them (implemented) or feeding them treats (not yet). Dogs that you've befriended can be sent to fetch things on command simply by pointing to a fetchable item in sight of the dog, at which point the dog will go after it and attempt to bring it back, whether or not you actually meant to tell them to fetch it. I'd like to let you change whether the dog will fetch/attack/trip/whatever your target based on what you say to them, but that'll come at a later date if/when more behaviors are added. For now, I'm keeping it simple to learn the new AI system properly and not overstretch.

There's still a fair bit to do before this is 100% ready, here's a quick list off the top of my head:

- [ ] Port normal dog behavior of eating snacks they find to the new system
- [ ] Properly segment the AI controller and behaviors so things are in the right place
- [ ] Make sure all edge-cases are appropriately handled like if a fetch target is deleted or moves to another z-level or whatever.

## Why It's Good For The Game
Making pets more lifelike and interesting than simply standing around waiting to get shot would go a long way towards improving immersion. Lots of possibilities for !fun! emergent gameplay like Ian intercepting a hand grenade thrown by a nuke op and chasing them down with it, or simply being a lazy warden and getting the brig dog to fetch you a shotgun from the next room without having to get it yourself.

## Changelog
:cl: Ryll/Shaps
add: Dogs have been given an upgrade to their AI! They'll now try fetching anything thrown in their sight and bringing it back to the thrower. You can also command a dog you've befriended to bring you something by pointing at the item while the dog can see it.
add: Yes, dogs will try to fetch hand grenades, and yes, they'll bring them back to whoever threw them if possible.
/:cl:

### CODE REVIEW

File: code/modules/mob/mob.dm [hunk 1]
```diff
@@ -578,6 +578,7 @@
 
 	point_at(A)
 
+	SEND_SIGNAL(src, COMSIG_MOB_POINTED, A)
```
Comment (Jared-Fogle) [inline code/modules/mob/mob.dm:581]: finally, damn daniel shoes support

**Comment by Timberpoes** [comment]
Converted to draft as per our contributor guidelines. When this is ready for full review, please un-draft it to let us know it's ready!

File: code/datums/ai/dog/dog_controller.dm [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+		current_movement_target = return_target
+		current_behaviors += GET_AI_BEHAVIOR(/datum/ai_behavior/dog_deliver)
+		blackboard[BB_DOG_DELIVERING] = TRUE
+		return
+
+	var/list/old_throw_listeners = blackboard[BB_DOG_THROW_LISTENERS]
+	var/list/new_throw_listeners = list()
+
+	// check around us for people who we can hear throw things for fetching
+	// (would probably be smart to make this run less frequently than every tick)
```
Comment (Qustinnus) [inline code/datums/ai/dog/dog_controller.dm:52]: theres a var for this now

Comment (Qustinnus) [reply code/datums/ai/dog/dog_controller.dm:52]: also why dont you just way for someone to say "Fetch" and then listen for that person to throw something?

File: code/datums/ai/dog/dog_behaviors.dm [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	if(in_range(living_pawn, fetch_thing))
+		finish_action(controller, TRUE)
+		return
+
+	finish_action(controller, FALSE)
+
+/datum/ai_behavior/dog_fetch/finish_action(datum/ai_controller/controller, success)
+	. = ..()
+
+	if(!success) //Don't try again on this item if we failed
```
Comment (Qustinnus) [inline code/datums/ai/dog/dog_behaviors.dm:25]: finish_action should be for clean-up, not for extra behavior, this should be in perform()

File: code/datums/ai/dog/dog_controller.dm [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+/datum/ai_controller/dog/PerformIdleBehavior(delta_time)
+	// todo: port dogs eating snacks
+	var/mob/living/living_pawn = pawn
+
+	if(DT_PROB(25, delta_time) && (living_pawn.mobility_flags & MOBILITY_MOVE) && isturf(living_pawn.loc) && !living_pawn.pulledby)
+		var/move_dir = pick(GLOB.alldirs)
+		living_pawn.Move(get_step(living_pawn, move_dir), move_dir)
+
+/// Someone we were listening to throws for has thrown something, start listening to the thrown item so we can see if we want to fetch it when it lands
+/datum/ai_controller/dog/proc/listened_throw(mob/living/carbon/carbon_thrower)
```
Comment (Qustinnus) [inline code/datums/ai/dog/dog_controller.dm:76]: make sure you have the things you need for a signal in here.

File: code/modules/mob/living/simple_animal/friendly/dog.dm [hunk 5]
```diff
@@ -19,6 +19,8 @@
 	can_be_held = TRUE
 	pet_bonus = TRUE
 	pet_bonus_emote = "woofs happily!"
+	ai_controller = /datum/ai_controller/dog
+	stop_automated_movement = TRUE
```
Comment (Qustinnus) [inline code/modules/mob/living/simple_animal/friendly/dog.dm:23]: You need to remove all the old AI behavior too.

Comment (Ryll-Ryll) [reply code/modules/mob/living/simple_animal/friendly/dog.dm:23]: Yeah, that's covered in my checklist of stuff to do

File: code/datums/ai/dog/dog_controller.dm [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+		UnregisterSignal(thrown_thing, list(COMSIG_PARENT_QDELETING, COMSIG_MOVABLE_THROW_LANDED))
+		return
+
+	current_movement_target = thrown_thing
+	blackboard[BB_DOG_FETCH_TARGET] = thrown_thing
+	blackboard[BB_DOG_FETCH_THROWER] = throwing_datum.thrower
+	current_behaviors += GET_AI_BEHAVIOR(/datum/ai_behavior/dog_fetch)
+
+/// Someone's interacting with us by hand, see if they're being nice or mean
+/datum/ai_controller/dog/proc/on_attack_hand(datum/source, mob/living/user)
```
Comment (Qustinnus) [inline code/datums/ai/dog/dog_controller.dm:101]: Are you removing the normal petting behavior as well?

Comment (Qustinnus) [reply code/datums/ai/dog/dog_controller.dm:101]: in simplemobs that is

Comment (Ryll-Ryll) [reply code/datums/ai/dog/dog_controller.dm:101]: Nope, this is simply in addition to petting

File: code/datums/ai/dog/dog_controller.dm (hunk 2 above)
Comment (Jared-Fogle) [reply code/datums/ai/dog/dog_controller.dm:52]: 1. Not intuitive
2. Less people would unknowingly throw grenades for Ian to bring them back.

**Comment by Zytolg** [comment]
They say that we're taking the soul away from the game.
This PR gives Ian a soul.
Ian/10 I'm begging for a speedmerge again

File: code/__DEFINES/ai/ai.dm [hunk 7]
```diff
@@ -60,3 +60,14 @@
 #define BB_VENDING_UNTILT_COOLDOWN "BB_vending_untilt_cooldown"
 #define BB_VENDING_BUSY_TILTING "BB_vending_busy_tilting"
 #define BB_VENDING_LAST_HIT_SUCCESFUL "BB_vending_last_hit_succesful"
+
+///Dog AI controller blackboard keys
+
+#define BB_FETCHING "BB_FETCHING"
```
Comment (Qustinnus) [inline code/__DEFINES/ai/ai.dm:66]: should this be in the dogs.dm?

**Comment by Ryll-Ryll** [comment]
This has been shelved while I've been working on JPS, but I'll return for this shortly

**Pull request closed without merging**